torch/distributed/_composable/checkpoint_activation.py

```diff
@@ -214,63 +62,39 @@ def checkpoint(module: nn.Module, *, use_reentrant: bool = True) -> nn.Module:
-                # run_function, we SHOULD actually stash the cuda state here.  Unfortunately,
-                # we have no way to anticipate this will happen before we run the function.)
-                checkpoint.state(module).had_cuda_in_fwd = False
-                if torch.cuda._initialized:
-                    checkpoint.state(module).had_cuda_in_fwd = True
-                    (
-                        checkpoint.state(module).fwd_gpu_devices,
-                        checkpoint.state(module).fwd_gpu_states,
-                    ) = get_device_states(*inputs)
 
-            else:
-                # The Holder object for each of the saved object is saved
-                # directly on the SavedVariable and is cleared when reset_data()
-                # is called on it. We MUST make sure that this is the only
-                # object having an owning reference to ensure that the Tensor
-                # stored in storage is deleted as soon as the corresponding
-                # SavedVariable data is cleared.
-                storage: WeakKeyDictionary = WeakKeyDictionary()
-                weak_holder_list: List[ReferenceType] = []
-                saved_tensor_hooks = torch.autograd.graph.saved_tensors_hooks(
-                    partial(_pack, weak_holder_list=weak_holder_list),
-                    partial(
-                        _unpack,
-                        storage=storage,
-                        weak_holder_list=weak_holder_list,
-                        module=module,
-                        inputs=inputs,
-                    ),
-                )
-                saved_tensor_hooks.__enter__()
-                checkpoint.state(module).saved_tensor_hooks = saved_tensor_hooks
+            def context_fns():
+                return nullcontext(), _no_hook(module)
+
+            checkpoint.state(
+                module
+            )._ac_generator = _checkpoint_without_reentrant_generator(
+                module,
+                True,
+                context_fns,
+                _DEFAULT_DETERMINISM_MODE,
+                False,
+                *inputs
+            )
+            next(checkpoint.state(module)._ac_generator)
 
     def forward_hook(module: nn.Module, inputs: Tuple[Any, ...], output: Any) -> Any:
         if checkpoint.state(module).enable_hook:
-            torch.set_grad_enabled(checkpoint.state(module).orig_grad_enabled)
-            if checkpoint.state(module).use_reentrant:
-                return _ModuleHookCheckpointFunction.apply(module, output, *inputs)
+            try:
+                next(checkpoint.state(module)._ac_generator)
+            except StopIteration:
+                pass
             else:
-                checkpoint.state(module).saved_tensor_hooks.__exit__()
-                checkpoint.state(module).saved_tensor_hooks = None
+                raise RuntimeError(
+                    "Expected non-reentrant activation checkpoint generator to be exhausted, but it was not!"
```

Comment:
What happens if we hit this error? What do we do?

Reply:
I'm not sure if this can happen other than an implementation change / bug where the AC generator can be `next`'ed more than twice. I guess we should tell the user to file a repro?

Reply:
Sounds good!